File: src/services/stats.ts
import { sql } from "kysely";
import { db } from "../db";
import { IndexStatus, JobStatus, type IndexJob, type Site } from "../db/types";
import { dayRange, daysAgo, todayIso } from "../lib/dates";

export async function recordSiteSnapshot(
  site: Pick<Site, "id">,
  submitted = 0,
  inspected = 0,
): Promise<void> {
  const day = todayIso();
  const rows = await db
    .selectFrom("urls")
    .select(["index_status", sql<number>`COUNT(id)`.as("total")])
    .where("site_id", "=", site.id)
    .where("is_active", "=", true)
    .groupBy("index_status")
    .execute();

  const counts = new Map(rows.map((row) => [row.index_status, Number(row.total)]));
  const totalUrls = [...counts.values()].reduce((sum, value) => sum + value, 0);
  const indexed = counts.get(IndexStatus.INDEXED) ?? 0;
  const notIndexed =
    (counts.get(IndexStatus.NOT_INDEXED) ?? 0) + (counts.get(IndexStatus.EXCLUDED) ?? 0);

  // Migawka dnia jest jedna, ale wywolan w ciagu dnia wiele - liczniki
  // zgloszen i inspekcji musza sie sumowac, a stany indeksowania nadpisywac.
  await sql`
    INSERT INTO site_stats (site_id, day, total_urls, indexed, not_indexed, submitted, inspected)
    VALUES (${site.id}, ${day}, ${totalUrls}, ${indexed}, ${notIndexed}, ${submitted}, ${inspected})
    ON DUPLICATE KEY UPDATE
      total_urls = VALUES(total_urls),
      indexed = VALUES(indexed),
      not_indexed = VALUES(not_indexed),
      submitted = submitted + VALUES(submitted),
      inspected = inspected + VALUES(inspected)
  `.execute(db);
}

export interface HistoryPoint {
  day: string;
  indexed: number;
  not_indexed: number;
  total: number;
}

export async function workspaceIndexingHistory(
  workspaceId: number,
  days = 30,
): Promise<HistoryPoint[]> {
  const start = todayIso(-(days - 1));
  const rows = await db
    .selectFrom("site_stats")
    .innerJoin("sites", "sites.id", "site_stats.site_id")
    .select([
      "site_stats.day as day",
      sql<number>`SUM(site_stats.indexed)`.as("indexed"),
      sql<number>`SUM(site_stats.not_indexed)`.as("not_indexed"),
      sql<number>`SUM(site_stats.total_urls)`.as("total"),
    ])
    .where("sites.workspace_id", "=", workspaceId)
    .where("site_stats.day", ">=", start)
    .groupBy("site_stats.day")
    .orderBy("site_stats.day")
    .execute();

  const known = new Map(
    rows.map((row) => [
      typeof row.day === "string" ? row.day : new Date(row.day).toISOString().slice(0, 10),
      {
        indexed: Number(row.indexed ?? 0),
        not_indexed: Number(row.not_indexed ?? 0),
        total: Number(row.total ?? 0),
      },
    ]),
  );

  // Dni bez migawki dziedzicza wartosci z poprzedniego dnia. Inaczej wykres
  // spadalby do zera zawsze, gdy panel nie mial nic do zrobienia.
  const series: HistoryPoint[] = [];
  let last = { indexed: 0, not_indexed: 0, total: 0 };
  for (const day of dayRange(days)) {
    last = known.get(day) ?? last;
    series.push({ day, ...last });
  }
  return series;
}

export interface RecentJob extends IndexJob {
  site_name: string;
}

export function recentJobs(workspaceId: number, limit = 15): Promise<RecentJob[]> {
  return db
    .selectFrom("index_jobs")
    .innerJoin("sites", "sites.id", "index_jobs.site_id")
    .selectAll("index_jobs")
    .select("sites.display_name as site_name")
    .where("sites.workspace_id", "=", workspaceId)
    .orderBy("index_jobs.created_at", "desc")
    .limit(limit)
    .execute() as Promise<RecentJob[]>;
}

export async function jobTotals(
  workspaceId: number,
  days = 30,
): Promise<Record<string, number>> {
  const rows = await db
    .selectFrom("index_jobs")
    .innerJoin("sites", "sites.id", "index_jobs.site_id")
    .select(["index_jobs.status as status", sql<number>`COUNT(index_jobs.id)`.as("total")])
    .where("sites.workspace_id", "=", workspaceId)
    .where("index_jobs.created_at", ">=", daysAgo(days))
    .groupBy("index_jobs.status")
    .execute();

  const totals: Record<string, number> = Object.fromEntries(
    Object.values(JobStatus).map((status) => [status, 0]),
  );
  for (const row of rows) totals[row.status] = Number(row.total);
  totals["all"] = Object.values(JobStatus).reduce((sum, s) => sum + (totals[s] ?? 0), 0);
  return totals;
}

File: app/services/stats.py
from __future__ import annotations

from datetime import date, datetime, timedelta, timezone

from sqlalchemy import func, select
from sqlalchemy.orm import Session

from app.models import IndexJob, IndexStatus, JobStatus, PageUrl, Site, SiteStat


def record_site_snapshot(db: Session, site: Site, submitted: int = 0, inspected: int = 0) -> SiteStat:
    day = datetime.now(timezone.utc).date()
    rows = db.execute(
        select(PageUrl.index_status, func.count(PageUrl.id))
        .where(PageUrl.site_id == site.id, PageUrl.is_active.is_(True))
        .group_by(PageUrl.index_status)
    ).all()
    counts = {status: count for status, count in rows}

    stat = db.scalar(select(SiteStat).where(SiteStat.site_id == site.id, SiteStat.day == day))
    if stat is None:
        stat = SiteStat(site_id=site.id, day=day)
        db.add(stat)

    stat.total_urls = sum(counts.values())
    stat.indexed = counts.get(IndexStatus.INDEXED, 0)
    stat.not_indexed = counts.get(IndexStatus.NOT_INDEXED, 0) + counts.get(IndexStatus.EXCLUDED, 0)
    stat.submitted = (stat.submitted or 0) + submitted
    stat.inspected = (stat.inspected or 0) + inspected
    db.flush()
    return stat


def workspace_indexing_history(db: Session, workspace_id: int, days: int = 30) -> list[dict]:
    start: date = datetime.now(timezone.utc).date() - timedelta(days=days - 1)
    rows = db.execute(
        select(
            SiteStat.day,
            func.sum(SiteStat.indexed),
            func.sum(SiteStat.not_indexed),
            func.sum(SiteStat.total_urls),
        )
        .join(Site, Site.id == SiteStat.site_id)
        .where(Site.workspace_id == workspace_id, SiteStat.day >= start)
        .group_by(SiteStat.day)
        .order_by(SiteStat.day)
    ).all()

    known = {
        day.isoformat() if hasattr(day, "isoformat") else str(day): {
            "indexed": int(indexed or 0),
            "not_indexed": int(not_indexed or 0),
            "total": int(total or 0),
        }
        for day, indexed, not_indexed, total in rows
    }

    series = []
    last = {"indexed": 0, "not_indexed": 0, "total": 0}
    for i in range(days):
        key = (start + timedelta(days=i)).isoformat()
        point = known.get(key, last)
        last = point
        series.append({"day": key, **point})
    return series


def recent_jobs(db: Session, workspace_id: int, limit: int = 15) -> list[IndexJob]:
    return list(
        db.scalars(
            select(IndexJob)
            .join(Site, Site.id == IndexJob.site_id)
            .where(Site.workspace_id == workspace_id)
            .order_by(IndexJob.created_at.desc())
            .limit(limit)
        )
    )


def job_totals(db: Session, workspace_id: int, days: int = 30) -> dict:
    start = datetime.now(timezone.utc).replace(tzinfo=None) - timedelta(days=days)
    rows = db.execute(
        select(IndexJob.status, func.count(IndexJob.id))
        .join(Site, Site.id == IndexJob.site_id)
        .where(Site.workspace_id == workspace_id, IndexJob.created_at >= start)
        .group_by(IndexJob.status)
    ).all()
    totals = {status.value: 0 for status in JobStatus}
    for status, count in rows:
        totals[status] = count
    totals["all"] = sum(totals[s.value] for s in JobStatus)
    return totals
